js-envs-test-kit/src/main/scala/org/scalajs/jsenv/test/RunTests.scala

```diff
@@ -0,0 +1,165 @@
+
+  @Test
+  def syntaxErrorTest: Unit = {
+    import kit._
+
+    """
+    {
+    """.fails()
+  }
+
+  @Test // Failed in Phantom - #2053
+  def utf8Test: Unit = {
+    import kit._
+
+    """
+    console.log("\u1234");
+    """ hasOutput "\u1234\n";
+  }
+
+  @Test
+  def allowScriptTags: Unit = {
+    import kit._
+
+    """
+    console.log("<script></script>");
+    """ hasOutput "<script></script>\n";
+  }
+
+  @Test
+  def jsExitsTest: Unit = {
+    assumeTrue(config.terminateVMJSCode.isDefined)
+
+    val run = kit.start(config.terminateVMJSCode.get, RunConfig())
+    try {
+      Await.result(run.future, Duration.Inf)
+    } finally {
+      run.close()
+    }
+  }
+
+  @Test // Node.js strips double percentage signs - #500
+  def percentageTest: Unit = {
+    import kit._
+
+    val counts = 1 to 15
+    val argcs  = 1 to 3
+    val strings = counts.map("%" * _)
+
+    val strlists = for {
+      count  <- argcs
+      string <- strings
+    } yield List.fill(count)(string)
+
+    val codes = for {
+      strlist <- strlists
+    } yield {
+      val args = strlist.map(s => s""""$s"""").mkString(", ")
+      s"console.log($args);\n"
+    }
+
+    val result = strlists.map(_.mkString(" ") + "\n").mkString("")
+
+    codes.mkString("").hasOutput(result)
+  }
+
+  @Test // Node.js console.log hack didn't allow to log non-Strings - #561
+  def nonStringTest: Unit = {
+    import kit._
+
+    """
+    console.log(1);
+    console.log(undefined);
+    console.log(null);
+    console.log({});
+    console.log([1,2]);
+    """ hasOutput
+    """|1
+       |undefined
+       |null
+       |[object Object]
+       |1,2
+    |""".stripMargin
+  }
+
+  @Test
+  def fastCloseTest: Unit = {
+    /* This test also tests a failure mode where the ExternalJSRun is still
+     * piping output while the client calls close.
+     */
+    val run = kit.start("", RunConfig())
+    run.close()
+
+    if (withCom)
+      Await.result(run.future, Duration.Inf)
+    else
+      Await.ready(run.future, Duration.Inf)
+  }
+
+  @Test
+  def multiCloseAfterTerminatedTest: Unit = {
+    val run = kit.start("", RunConfig())
+    run.close()
+
+    Await.ready(run.future, Duration.Inf)
+
+    // Should be noops (and not fail).
+    run.close()
+    run.close()
+    run.close()
+  }
+
+  @Test
+  def noThrowOnBadFileTest: Unit = {
+    val badFile = new VirtualJSFile {
+      def path: String = ???
+      def exists: Boolean = ???
+      def content: String = ???
+    }
+
+    // `start` may not throw but must fail asynchronously
+    val run = kit.start(badFile, RunConfig())
+    try {
+      Await.ready(run.future, Duration.Inf)
+      assertTrue("Bad file should have made run fail",
+        run.future.value.get.isFailure)
+    } finally {
+      run.close()
+    }
+  }
+
+  @Test(expected = classOf[RunConfig.FailedValidationForTestException])
+  def ensureValidate: Unit = {
+    val cfg = RunConfig().withFailValidationForTest(true)
+    kit.start("", cfg).close()
```

Comment:
So that's what the dummy `private[jsenv]` exception type and field in `RunConfig` are for? That's pretty twisted. (also the error message that JUnit will be super confusing and not very actionable :-s)

Reply:
Yes. I can add a big comment explaining what this does, but I don't know how to do this better so we can still test it. I think it is important that we test this.

Reply:
Neither do I--and I tried finding a better way--, or I would have suggested something :(
And yes, it's probably important, because otherwise custom JSEnv authors will never call that thing.

Reply:
I have
- removed the custom exception (it is not necessary for the test to be effective)
- improved the name of the option
- added some more doc / comments